ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/HeartbeatHandler.java

```diff
@@ -0,0 +1,332 @@
+    this.serviceContext = requireNonNull(serviceContext, "serviceContext");
+    this.configs = requireNonNull(configs, "configuration parameters");
+    this.scheduledExecutorService = Executors.newScheduledThreadPool(configs.threadPoolSize);
+    this.receivedHeartbeats = new ConcurrentHashMap<>();
+    this.aliveHosts = new ConcurrentHashMap<>();
+    this.remoteHostsURI = new ConcurrentHashMap<>();
+    this.remoteHostsInfo = new ConcurrentHashMap<>();
+    this.localHostInfo = new HostInfo(localServerURI.getHost(), localServerURI.getPort());
+    try {
+      this.localURL = localServerURI.toURL();
+    } catch (final Exception e) {
+      throw new IllegalStateException("Failed to convert local host URI to URL."
+                                          + " local host URL: " + localServerURI);
+    }
+    this.aliveHosts.putIfAbsent(localHostInfo.toString(), true);
+  }
+
+  void scheduleHeartbeatRelatedTasks() {
+    SendHeartbeatTask sendTask = new SendHeartbeatTask();
+    scheduledExecutorService.scheduleAtFixedRate(sendTask, 0, configs.heartbeatSendIntervalMs,
+                                                 TimeUnit.MILLISECONDS);
+    CheckHeartbeatTask processTask = new CheckHeartbeatTask();
+    scheduledExecutorService.scheduleAtFixedRate(processTask, 0, configs.heartbeatCheckIntervalMs,
+                                                 TimeUnit.MILLISECONDS);
+    DiscoverClusterTask discoverTask = new DiscoverClusterTask();
+    scheduledExecutorService.scheduleAtFixedRate(discoverTask, 0, configs.discoverClusterIntervalMs,
+                                                 TimeUnit.MILLISECONDS);
+
+  }
+
+  void shutDown() {
+    scheduledExecutorService.shutdown();
+  }
+
+  public ConcurrentHashMap<HostInfo, TreeMap<Long, HeartbeatInfo>> getReceivedHeartbeats() {
+    return receivedHeartbeats;
+  }
+
+  ConcurrentHashMap<URI, Boolean> getRemoteHostsURI() {
+    return remoteHostsURI;
+  }
+
+  public Map<String, Boolean> getAliveHosts() {
+    return aliveHosts;
+  }
+
+  class CheckHeartbeatTask implements Runnable {
+
+    @Override
+    public void run() {
+      long now = System.currentTimeMillis();
+      long windowStart = now - configs.heartbeatWindowMs;
+      processHeartbeats(windowStart, now);
+    }
+
+    @VisibleForTesting
+    void runWithWindow(long windowStart, long windowEnd) {
+      processHeartbeats(windowStart, windowEnd);
+    }
+
+    private void processHeartbeats(final long windowStart, final long windowEnd) {
+      if (receivedHeartbeats.isEmpty()) {
+
+        remoteHostsInfo.forEach((host, bool) ->
+                                    aliveHosts.compute(host.toString(), (key, val) -> false));
+      }
+      for (Entry<HostInfo, TreeMap<Long, HeartbeatInfo>> entry : receivedHeartbeats.entrySet()) {
+        HostInfo host = entry.getKey();
+        TreeMap<Long, HeartbeatInfo> heartbeats = entry.getValue();
+        synchronized (heartbeats) {
```

Comment:
instead of synchronizing, it might make sense to use `ConcurrentNavigableMap` (implementation is `ConcurrentSkipListMap`)

Reply:
We discussed this internally before.. Decided to keep simple synchronized for now, since reasoning about execution on concurrent data structure is non-trivial and the performance needs here may not strictly warrant them. We can swap them out later as well.

Reply:
I agree that performance isn't critical - I just thought it would be easier to reason about the concurrency model if the produce/consume was split. I'm going to close out this (we can continue conversation on https://github.com/confluentinc/ksql/pull/4173#discussion_r360136881